metrics/resource_view.go

```diff
@@ -0,0 +1,294 @@
+func RegisterResourceView(views ...*view.View) error {
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	errors := make([]error, 0, len(allMeters.meters))
+
+	for _, meter := range allMeters.meters {
+		if err := meter.m.Register(views...); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return errors[0] // The first error is as good as any
+	}
+
+	resourceViews.views = append(resourceViews.views, views...)
+	return nil
+}
+
+func UnregisterResourceView(views ...*view.View) {
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		meter.m.Unregister(views...)
+	}
+
+	j := 0
+	for _, view := range resourceViews.views {
+		toRemove := false
+		for _, viewToRemove := range views {
+			if view == viewToRemove {
+				toRemove = true
+				break
+			}
+		}
+		if !toRemove {
+			resourceViews.views[j] = view
+			j++
+		}
+	}
+	resourceViews.views = resourceViews.views[:j]
+}
+
+func setFactory(f func(*resource.Resource) (view.Exporter, error)) error {
+	if f == nil {
+		return errors.New("do not setFactory(nil)")
+	}
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	allMeters.factory = f
+
+	errs := make([]error, len(allMeters.meters))
+
+	for r, meter := range allMeters.meters {
+		e, err := f(resourceFromString(r))
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		meter.m.UnregisterExporter(meter.e)
+		meter.m.RegisterExporter(e)
+		meter.e = e
+	}
+	if len(errs) > 0 {
+		return errs[0]
+	}
+	return nil
+}
+
+func flushResourceExporters() {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		flushGivenExporter(meter.e)
+	}
+}
+
+func meterExporterForResource(r *resource.Resource) *meterExporter {
+	key, ok := allMeters.resourceToKey[r]
+	if !ok {
+		key = resourceAsString(r)
+		allMeters.resourceToKey[r] = key
+	}
+
+	mE := allMeters.meters[key]
+	if mE == nil {
+		mE = &meterExporter{}
+		allMeters.meters[key] = mE
+	}
+	if mE.o != nil {
+		return mE
+	}
+	mE.m = view.NewMeter()
+	mE.m.SetResource(r)
+	mE.m.Start()
+	mE.m.Register(resourceViews.views...)
+	/*
+		// Prometheus's default collector uses the global metricproducer to read *all* meters.
+		// This confuses the default prometheus adapter, so remove these from the global export.
+		metricproducer.GlobalManager().DeleteProducer(mE.m.(metricproducer.Producer))
+	*/
+	mE.o = stats.WithRecorder(mE.m)
+	allMeters.meters[key] = mE
+	return mE
+}
+
+// optionForResource finds or creates a stats.Option indicating which meter to record to.
+func optionForResource(r *resource.Resource) (stats.Options, error) {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	mE := meterExporterForResource(r)
+	if mE == nil {
+		return nil, fmt.Errorf("unexpectedly failed lookup for resource %v", r)
```

Comment:
```suggestion
		return nil, fmt.Errorf("failed lookup for resource %v", r)
```
I'd presume expected failures are in tests, hopefully :D

Reply:
Currently, `meterExporterForResource` never return `nil`, so this was purely defensive.